table_engine/src/partition/rule/key.rs

```diff
@@ -401,6 +401,39 @@ mod tests {
         assert_eq!(total_bytes, expect_total_bytes);
     }
 
+    #[test]
+    fn test_compute_partition_for_inserted_row_with_empty_string() {
+        let partition_num = 16;
+        let key_rule = KeyRule {
+            typed_key_columns: vec![ColumnWithType::new("col1".to_string(), DatumKind::String)],
+            partition_num,
+        };
+
+        let datums = vec![
+            Datum::Int32(1),
+            Datum::String(StringBytes::copy_from_str("")),
+            Datum::String(StringBytes::copy_from_str("xtest")),
+            Datum::Int64(84),
+            Datum::Null,
+        ];
+        let row = Row::from_datums(datums.clone());
+        let defined_idxs = vec![1, 2, 3, 4];
+
+        // Actual
+        let actual = key_rule.compute_partition_for_inserted_row(&row, &defined_idxs);
+
+        // Expected
+        let mut buf = BytesMut::new();
+        buf.clear();
+        buf.put_slice(&datums[1].to_bytes());
+        buf.put_slice(&datums[2].to_bytes());
+        buf.put_slice(&datums[3].to_bytes());
+        buf.put_slice(&datums[4].to_bytes());
+        let expected = (hash64(&buf[..]) % (partition_num as u64)) as usize;
+
+        assert_eq!(actual, expected);
```

Comment:
I suggest we use constants(like `assert_eq!(actual, 123)`) to do check, so we can ensure future updates won't break any more.